### PR TITLE
add eviction subresource - for #127

### PR DESCRIPTION
**Pull request by clux**
might rebase against the big hyper pr, but it was fairly trivial to get this to work at least.

### CODE REVIEW

**Comment by chroche** [comment]
Wow that was fast! :-)

Thanks a lot for that

File: kube/src/api/subresource.rs [hunk 1]
```diff
@@ -221,6 +222,84 @@ where
...
+    use k8s_openapi::api::core::v1 as corev1;
+    let r = Resource::namespaced::<corev1::Pod>("ns");
+    let mut pp = PostParams::default();
+    pp.dry_run = true;
+    let req = r.evict("foo", &pp, vec![]).unwrap();
+    assert_eq!(req.uri(), "/api/v1/namespaces/ns/pods/foo/eviction?&dryRun=All");
+}
+
+/// Marker trait for objects that can be evicted
+pub trait EvictionObject {}
```
Comment (teozkr) [inline kube/src/api/subresource.rs:287]: Nit: `Evictable`, maybe? `EvictionObject` sounds like it's the reified object created when evicting something.

File: kube/src/api/subresource.rs [hunk 2]
```diff
@@ -221,6 +222,84 @@ where
...
+
+use k8s_openapi::apimachinery::pkg::apis::meta::v1::ObjectMeta;
+/// Eviction body
+#[derive(Default, Clone, Serialize)]
+pub struct Eviction {
+    /// How the eviction should occur
+    pub delete_options: Option<DeleteParams>,
+    /// Metadata describing the pod being evicted
+    /// we somehow need this despite providing the name of the pod..
+    pub metadata: ObjectMeta,
```
Comment (teozkr) [inline kube/src/api/subresource.rs:237]: That feels like a pretty big footgun.. Maybe `Api::evict` could take the `name` from the `Eviction` (or the reverse)?

**Review by teozkr — APPROVED** [review]

File: kube/src/api/subresource.rs (hunk 1 above)
Comment (clux) [reply kube/src/api/subresource.rs:287]: Hm, I thought we had a convention, on noun-ing. But I am wrong.
Maybe we should just go full on on the conjugated form and rename all the traits?

i.e.
```rust
pub trait Loggable
pub trait Evictable
pub trait Executable
pub trait Atttachable
```

File: kube/src/api/subresource.rs (hunk 2 above)
Comment (clux) [reply kube/src/api/subresource.rs:237]: yeah, we could just pave over that interface, and take a (DeleteParams, PostParams, name) and only have the Eviction struct be internal. I don't see what good the metadata would do if it's tied to a named pod anyway.

Comment (clux) [reply kube/src/api/subresource.rs]: Made an EvictParams object containing the two params objects, and hid Eviction in the body of `Resource::evict`

File: kube/src/api/subresource.rs (hunk 1 above)
Comment (kazk) [reply kube/src/api/subresource.rs]: I was thinking EvictableObject like AttachableObject.
Logging and Executing sounds fine to me, but not against making them consistent.
Naming is hard.

Comment (teozkr) [reply kube/src/api/subresource.rs]: Sounds good to me. I have been thinking about proposing that for *all* resource actions at some point (for example: `ComponentStatus` is read-only: https://v1-18.docs.kubernetes.io/docs/reference/generated/kubernetes-api/v1.18/#componentstatus-v1-core).

Comment (clux) [reply kube/src/api/subresource.rs]: I'll make a follow up PR for standardising. Sounds like we are OK with the conjugation as a convention.

**Comment by clux** [comment]
false negative on kind ci again, need to fix that at some point.
circleci took almost 30 minutes to run `cargo_test` on everything.. hopefully that gets better once the caching kicks in.

**Pull request merged**